### PR TITLE
Add flag to log all output to a file

### PR DESCRIPTION
**Pull request by font**
This is helpful for running cosign within a container when you do not have access to a shell to redirect output to a file. For example, using the cosign container within CI systems e.g. as a task in Tekton, makes redirecting all output to a file very useful for easier post-processing to analyze the results.

### CODE REVIEW

File: cmd/cosign/main.go [hunk 1]
```diff
@@ -47,22 +49,39 @@ func main() {
...
+			fmt.Fprintf(os.Stderr, "error: %v\n", errors.Wrapf(err, "Error creating output file %s", *outputFilename))
+			os.Exit(1)
 		}
-		fmt.Fprintf(os.Stderr, "error: %v\n", err)
-		os.Exit(1)
+		stdout := os.Stdout
+		stderr := os.Stderr
+		defer func() {
+			os.Stdout = stdout
+			os.Stderr = stderr
```
Comment (dlorenc) [inline cmd/cosign/main.go:65]: Do you think it makes sense to always capture both stdout and stderr? In general I've tried to be careful and put actual "output" in stdout, and user-centered "messages" to stderr.

File: cmd/cosign/main.go [hunk 2]
```diff
@@ -47,22 +49,39 @@ func main() {
...
-		}
-		fmt.Fprintf(os.Stderr, "error: %v\n", err)
-		os.Exit(1)
+		printErrAndExit(err)
+	}
+}
+
+func printErrAndExit(err error) {
+	if *verbose {
+		fmt.Print("verbose!")
```
Comment (dlorenc) [inline cmd/cosign/main.go:83]: I think this can get deleted :) It was a copy/paste error IIRC.

**Comment by dlorenc** [comment]
Big +1 to this overall, one question on stderr

File: cmd/cosign/main.go (hunk 1 above)
Comment (font) [reply cmd/cosign/main.go:65]: I'm not sure I'm clear on the distinction. For example, https://github.com/SigStore/cosign/blob/2befb2893ee5ac5e697ed09fd4c146eef98a6ab7/cmd/cosign/cli/verify.go#L127 contains print statements to `stderr` for results of verifying claims, tlog entry, etc., while logging error messages to `stdout`. I think I would expect the opposite in that case and so I wasn't able to find a consistent pattern throughout to understand a distinction. As a result I found needing to see both stdout and stderr messages. But that's not necessarily a reason to log both to a file, and in principle I agree and think it's useful to separate them or at least duplicate them if nothing else.

We can just log to `stdout` for now (leave `stderr` untouched), file an issue to make `stdout`/`stderr` logging convention consistent and perhaps we'll need to consider revamping the entire print approach to use a more abstracted logging mechanism. WDYT?

File: cmd/cosign/main.go (hunk 2 above)
Comment (font) [reply cmd/cosign/main.go:83]: I didn't know if it was part of more elaborate plans. :smile: I'll remove in the next push and remove the flag as well.

File: cmd/cosign/main.go (hunk 1 above)
Comment (dlorenc) [reply cmd/cosign/main.go:65]: > We can just log to `stdout` for now (leave `stderr` untouched), file an issue to make `stdout`/`stderr` logging convention consistent and perhaps we'll need to consider revamping the entire print approach to use a more abstracted logging mechanism. WDYT?

Perfect! I *think* that code you linked is actually working as I intended it to, but I admit the intention isn't very clear and could probably use improvements.

**Review by dlorenc — DISMISSED** [review]

File: cmd/cosign/main.go (hunk 1 above)
Comment (font) [reply cmd/cosign/main.go]: Cool! Filed https://github.com/sigstore/cosign/issues/263.

**Comment by font** [comment]
@dlorenc One more review needed - pushed an update to the flag usage.

**Review by dlorenc — APPROVED** [review]

**Pull request merged**